Sources/PostgresNIO/New/Connection State Machine/RowStreamStateMachine.swift

```diff
@@ -0,0 +1,163 @@
+import NIOCore
+
+/// A sub state for receiving data rows. Stores whether the consumer has either signaled demand and whether the
+/// channel has issued `read` events.
+///
+/// This should be used as a SubStateMachine in QuerySubStateMachines.
+struct RowStreamStateMachine {
+    
+    enum Action {
+        case read
+        case wait
+    }
+    
+    private enum State {
+        /// The state machines expects further writes to `channelRead`. The writes are appended to the buffer.
+        case waitingForRows(CircularBuffer<PSQLBackendMessage.DataRow>)
+        /// The state machines expects a call to `demandMoreResponseBodyParts` or `read`. The buffer is
+        /// empty. It is preserved for performance reasons.
+        case waitingForReadOrDemand(CircularBuffer<PSQLBackendMessage.DataRow>)
+        /// The state machines expects a call to `read`. The buffer is empty. It is preserved for performance reasons.
+        case waitingForRead(CircularBuffer<PSQLBackendMessage.DataRow>)
+        /// The state machines expects a call to `demandMoreResponseBodyParts`. The buffer is empty. It is
+        /// preserved for performance reasons.
+        case waitingForDemand(CircularBuffer<PSQLBackendMessage.DataRow>)
+
+        case modifying
+    }
+
+    private var state: State
+
+    init() {
+        self.state = .waitingForRows(CircularBuffer(initialCapacity: 32))
+    }
+
+    mutating func receivedRow(_ newRow: PSQLBackendMessage.DataRow) {
+        switch self.state {
+        case .waitingForRows(var buffer):
+            self.state = .modifying
+            buffer.append(newRow)
+            self.state = .waitingForRows(buffer)
+            
+        // For all the following cases, please note:
+        // Normally these code paths should never be hit. However there is one way to trigger
+        // this:
+        //
+        // If the server decides to close a connection, NIO will forward all outstanding
+        // `channelRead`s without waiting for a next `context.read` call. For this reason we might
+        // receive new rows, when we don't expect them here.
+        case .waitingForRead(var buffer):
+            self.state = .modifying
+            buffer.append(newRow)
+            self.state = .waitingForRows(buffer)
+            
+        case .waitingForDemand(var buffer):
+            self.state = .modifying
+            buffer.append(newRow)
+            self.state = .waitingForRows(buffer)
+            
+        case .waitingForReadOrDemand(var buffer):
+            self.state = .modifying
+            buffer.append(newRow)
+            self.state = .waitingForRows(buffer)
```

Comment:
Feel free to "tweak" my "code style" (generous to call it that, I agree) as you find apropos, I mostly just care about the code de-duplication.
```suggestion
        case .waitingForRows(var buffer),
             // For all the following cases, please note:
             // Normally these code paths should never be hit. However there is one way to trigger
             // this:
             //
             // If the server decides to close a connection, NIO will forward all outstanding
             // `channelRead`s without waiting for a next `context.read` call. For this reason we might
             // receive new rows, when we don't expect them here.
             .waitingForRead(var buffer),
             .waitingForDemand(var buffer),
             .waitingForReadOrDemand(var buffer):
            self.state = .modifying
            buffer.append(newRow)
            self.state = .waitingForRows(buffer
```